### PR TITLE
Check CUDA out of memory

### PR DESCRIPTION
**Pull request by thodkatz**
In conjunction with https://github.com/ilastik/ilastik/pull/2891, eventually our goal would be to find the maximum possible tensor shape that can fit to the GPU memory for a simple forward pass. 

To accomplish this, tiktorch should provide clients the functionality of doing the memory prompting to the GPU.

Two procedures have been added:
- `IsCudaOutOfMemory(shape) -> bool`, checks for a given shape if it fits to the GPU memory
- `MaxCudaMemoryShape(minShape, maxShape, step) -> shape`, for a range of [minShape, maxShape] and valid increments of step, returns the maximum shape that fits to the GPU memory.

Exposing the first one, could be still valuable if we have a good guess of a shape, since the second one can be very time consuming, depending on the range.

The implementation of prompting the GPU has been inspired as well by [plant-seg](https://github.com/kreshuklab/plant-seg/blob/6c90c750469b22efee1e49716b4e041ef91683de/plantseg/predictions/functional/array_predictor.py#L79).

### CODE REVIEW

**Review by k-dominik — COMMENTED** [review]
Awesome! Great that you got into tiktorch so quickly!

I left some minor comments and a question about how we want to see the interface for this functionality (basically ints vs namedints)...

File: proto/inference.proto [hunk 1]
```diff
@@ -131,6 +135,26 @@ message Tensor {
   repeated NamedInt shape = 3;
 }
 
+message IsCudaOutOfMemoryRequest {
+  string modelSessionId = 1;
+  repeated uint32 shapeReference = 2;
```
[the remark this answers is not in the document]
Comment (thodkatz) [reply proto/inference.proto:140]: Yep, I totally agree :)

Also, I was thinking to change the API a bit. Currently we have `shapeReference`, and `maxShapeReference`. That means that `shapeReference` must be compatible with the shape provided by the model with `modelSessionId`, and `maxShapeReference` must be also compatible, so we should validate them based on the model.

So I would like to get the step from the model, and expose a variable `stepMultiplier`, that controls how far we increment with step shape e.g. if `step = [1,1, 2, 2]` and `stepMultiplier = 2`, then `finalStep = [2, 2, 4, 4]`, and for `maximumShapeReference`, we don't need the explicit shape, but again a number to indicate the range of increments (`numIncrements`) using the `finalStep` e.g. if `minShapeReference = [1, 1, 256, 256]` and `finalStep=[0, 0, 16, 16]`, and `numIncrements=2], then the range we test the shapes would be [1, 1, 256, 256], [1, 1, 272, 272], [1,1,288, 288]`.

We need also to check if `minShapeReference` is compatible with the model. But again that implies that one of the input shapes of the model is parameterized so we can have this whole workflow. So we could again provide the `minShapeReference` and check if exists in the available `inputShapes`. Or we could expose a key of the `inputShapes`, instead of an explicit `minShapeReference`. That way we can resolve a use case when we have two different `inputShapes` with the same `minShape`. On the other hand, exposing the key, the index of the desired `inputShape`, can be still error prone, because not all `inputShapes` are parameterized.

That is braindump, sorry for that, I am just trying to figure out, how maybe it is more convenient for the client. What do you think :)

File: tiktorch/server/grpc/inference_servicer.py [hunk 2]
```diff
@@ -100,6 +106,67 @@ def Predict(self, request: inference_pb2.PredictRequest, context) -> inference_p
         pb_tensors = [converters.xarray_to_pb_tensor(res_tensor) for res_tensor in res]
         return inference_pb2.PredictResponse(tensors=pb_tensors)
 
+    def MaxCudaMemoryShape(
```
[the remark this answers is not in the document]
Comment (thodkatz) [reply tiktorch/server/grpc/inference_servicer.py:109]: sorry, what do you mean with code path

File: tiktorch/server/grpc/inference_servicer.py [hunk 3]
```diff
@@ -100,6 +106,67 @@ def Predict(self, request: inference_pb2.PredictRequest, context) -> inference_p
...
+                print(f"Using shape {shape} will cause out of memory.")
+            else:
+                raise
+        finally:
+            gc.collect()  # attempt to explicitly deallocate memory
+            torch.cuda.empty_cache()
+        return is_out_of_memory
+
+    def _is_gpu(self) -> bool:
+        return torch.cuda.is_available()
```
[the remark this answers is not in the document]
Comment (thodkatz) [reply tiktorch/server/grpc/inference_servicer.py:168]: yep, good catch :) I will try to resolve it with the device pool, thanks :)

**Comment by thodkatz** [comment]
hey @k-dominik :)

I have attempted to refactor the design of the `ModelInfo` and the concept of shapes.

The previous design of the `ModelInfo` was coupled to the representation of the `ModelSession` message used for the grpc communication. The new design creates a more feature-rich interface to work easily with the concept of shape, and the compression of the `ModelInfo` needed to be transferred by the server is localized.

Do you think that the design is on the right track?

File: tiktorch/server/session/process.py [hunk 4]
```diff
@@ -6,25 +8,234 @@
...
+    def axes(self) -> str:
+        return self.offset.axes
+
+
+class ParameterizedShape:
+    def __init__(self, min_shape: AxisWithValue, steps: AxisWithValue):
+        self._min_shape = min_shape
+        self._steps = steps
+        assert self._min_shape.is_same_axes(self._steps)
+        assert all(step == 0 for axis, step in steps if axis not in AxisWithValue.SPATIAL_AXES)  # todo: ?
```
Comment (thodkatz) [inline tiktorch/server/session/process.py:143]: I am not sure if this is valid? Does it make sense to allow non zero step values for non spatial axes increments?

File: tiktorch/server/grpc/inference_servicer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (k-dominik) [reply tiktorch/server/grpc/inference_servicer.py:96]: let's say I'm running on cpu, then this whole thing doesn't make much sense, and code assumes that one is on gpu - so there should be a code path if a caller calls this function, but session is using cpu?

**Review by k-dominik — COMMENTED** [review]
For the refactor of `ModelInfo`; In general I like the idea of having something more usable directly :), good idea!
With the implementation details for the "rich" nicer to use objects I left comments that mostly leaned towards subclassing `dict`, which are probably outdated as I'm thinking.... actually, what do you think about investigating if `bioimageio.spec` classes can be reused for this purpose. I think, at least for me, any removed layer of transformation would reduce cognitive load. On the other hand this would bind us to `bioimageio.spec`, at least for the metadata. But since that's the format we're supporting (and the only one) I don't see much of a problem.
Of course then I immediately ask myself why even do the dance and not parse the model on the client also, and be done with it (and we already do parse the spec on the client(ilastik) side.)...

Also the `ModelInfo` (and related convenience classes) would need a very heavy refactor when updating to 0.5 of the spec.

idk, what do you think @thodkatz?

**Comment by thodkatz** [comment]
Since the [PR](https://github.com/ilastik/tiktorch/pull/215) of removing the ModelInfo was merged, I have rebased this PR.

- I have removed some leftover code from the previous ModelInfo interface.
- I have addressed the comments regarding detection of gpu, with adding a `deviceId` field to the cuda memory requests and using the function [check_gpu_exists](https://github.com/ilastik/tiktorch/pull/213/files#diff-31a22ff2b156f62383aa4c24cb4bf40cd11cf238819548c8c29dc51e7dd0d4a0R171).
- Additional tests for invalid request of the `MaxCudaMemoryShape`.
- Since we are going to rely to bioimageio spec classes, I have removed any fancy objects, mentioned in the [refactoring state](https://github.com/ilastik/tiktorch/pull/213#issuecomment-2271220119).

File: tiktorch/converters.py [hunk 5]
```diff
@@ -1,38 +1,129 @@
...
-NamedShape = List[NamedInt]
-NamedVec = List[NamedFloat]
+NamedShape = Dict[str, int]
+NamedVec = Dict[str, float]
+
+
+def get_axes_with_size(axes: Tuple[str, ...], shape: Tuple[int, ...]) -> NamedShape:
+    if len(axes) != len(shape):
+        raise ValueError(f"{axes} and {shape} incompatible length. It should be equal")
+    InputTensorValidator.is_shape(shape)
```
Comment (thodkatz) [inline tiktorch/converters.py:21]: oops, this check should actually do something

File: tiktorch/converters.py [hunk 6]
```diff
@@ -1,38 +1,129 @@
...
+        return specs[0]
+
+    def _check_spec_exists(self, tensor_id: str):
+        spec_names = [spec.name for spec in self._input_specs]
+        if tensor_id not in spec_names:
+            raise ValueError(f"Spec {tensor_id} doesn't exist for specs {spec_names}")
+
+    def _check_shape_explicit(self, spec: nodes.InputTensor, tensor_shape: NamedShape):
+        assert self.is_shape_explicit(spec)
+        reference_shape = {name: size for name, size in zip(spec.axes, spec.shape)}
```
Comment (thodkatz) [inline tiktorch/converters.py:59]: we could use the `get_axes_with_size`

File: tiktorch/converters.py [hunk 7]
```diff
@@ -1,38 +1,129 @@
...
+        self.check_same_axes(reference_shape, tensor_shape)
+        if reference_shape != tensor_shape:
+            raise ValueError(f"Incompatible shapes found {tensor_shape}, expected {reference_shape}")
+
+    def _check_shape_parameterized(self, spec: nodes.InputTensor, tensor_shape: NamedShape):
+        assert isinstance(spec.shape, ParametrizedInputShape)
+        if not self.is_shape(tensor_shape.values()):
```
Comment (thodkatz) [inline tiktorch/converters.py:66]: if this is part of the `get_axes_with_size`, where we assume a valid named shape, meaning a map of the names of the axes, along with their size (natural number), `Dict[str, int]`, then we can remove this check from here, and integrate it into the `get_axes_with_size`.

**Comment by thodkatz** [comment]
Now that I think about it, we can improve the design:

What we actually need is just a validated xr.DataArray, and not really plain shapes. The xr.DataArray can be convenient since it contains already the logic of `dims`, that should represent the model axes. So we can utilize the `Sample` to create only validated `Sample`s. That way we keep the validation in one place. Currently, there is no way to infer if a sample is actually validate or not, and we don't valid shapes, but valid tensors.

File: tests/test_server/test_grpc/test_inference_servicer.py [hunk 8]
```diff
@@ -232,3 +242,119 @@ def test_call_predict_tf(self, grpc_stub, bioimageio_dummy_tensorflow_model_byte
         assert len(res.tensors) == 1
         assert res.tensors[0].tensorId == output_tensor_id
         assert_array_equal(expected, converters.pb_tensor_to_numpy(res.tensors[0]))
+
+
+class TestCudaMemory:
+    MAX_SHAPE = (1, 1, 10, 10)
+    AXES = ("b", "c", "y", "x")
```
Comment (k-dominik) [inline tests/test_server/test_grpc/test_inference_servicer.py:249]: maybe note that these values relate to the model being used - otherwise it's a bit hard to follow how `MAX_SHAPE` is enforced (i was first looking for some monkeypatching somewhere).

**Review by k-dominik — COMMENTED** [review]
Having this functionality is really great! I can already see us using gpus more efficiently!

In general it would be great to add docstrings to methods, especially if they are some sort of API. Take `IsCudaOutOfMemory` as an example. Given the name I'd assume that one can query if the current device is out of memory. But is looks like this is intended to check whether a given tensor with a given shape would fit.

Also at least somewhere the scope of what tiktorch supports in terms of models should be noted. When I read your code I think you assume single input tensors (the out of memory functionality should also add any additional tensors the model expects, with default sizes I guess, otherwise there will be errors thrown), do you think that limitation would be hard to lift?

Currently the out of memory tests will also not run on mac due to the model having pytorch state-dict weights (there is a [bug](https://github.com/ilastik/ilastik/issues/2827) currently, that prevents this from working). An easy workaround would be switching to torchscript. I think in current bioimage.spec/core this bug was fixed, but we're not there yet :). (Maybe we should add CI on a different platform, too... not in this PR ;))

I also noticed that there aren't any tests for `InputTensorValidator` - which would be great to have for completeness.

**Comment by thodkatz** [comment]
Great! Thank you very much @k-dominik for the review once more :) 

Totally agree with the statements, I will attempt to resolve them. Could you please also have a look on this comment https://github.com/ilastik/tiktorch/pull/213#issuecomment-2298563271, I think that a little bit of refactoring will remove some redundant checks, and it will be more readable as well :)

**Comment by thodkatz** [comment]
>Currently the out of memory tests will also not run on mac due to the model having pytorch state-dict weights (there is a https://github.com/ilastik/ilastik/issues/2827 currently, that prevents this from working). An easy workaround would be switching to torchscript. I think in current bioimage.spec/core this bug was fixed, but we're not there yet :). (Maybe we should add CI on a different platform, too... not in this PR ;))

Regarding this, actually without knowing it, I have attempted to fix the weight convresion functionality of the bioimage core in this [PR](https://github.com/bioimage-io/core-bioimage-io-python/pull/411)! But I think that this should indeed be handled by the core, and tiktorch shouldn't know about the weights format. But yeah let's keep the CI on another PR :)

>Also at least somewhere the scope of what tiktorch supports in terms of models should be noted. When I read your code I think you assume single input tensors (the out of memory functionality should also add any additional tensors the model expects, with default sizes I guess, otherwise there will be errors thrown), do you think that limitation would be hard to lift?

Yep you are right! I think it is time for the concept of `Sample` even from the proto buff perspective, so everything will be consistent with spec too :) I would assume that the client should provide for a Sample to check the out of memory functionality, so we don't really have to guess what we should do with other tensors if not specified. Although the sample increases the complexity a lot, because if you have for example two inputs with parameterized shape, and I want to test a range of shapes, then should I create a grid of all the possible combinations, should the shapes be updated synchronously?

**Comment by k-dominik** [comment]
> > Also at least somewhere the scope of what tiktorch supports in terms of models should be noted. When I read your code I think you assume single input tensors (the out of memory functionality should also add any additional tensors the model expects, with default sizes I guess, otherwise there will be errors thrown), do you think that limitation would be hard to lift?
> 
> Yep you are right! I think it is time for the concept of `Sample` even from the proto buff perspective, so everything will be consistent with spec too :) I would assume that the client should provide a `Sample` to check the out of memory functionality, so we don't really have to guess what we should do with other tensors if not specified. Although the `Sample` increases the complexity a lot, because if you have for example two inputs with parameterized shape, and I want to test a range of shapes, then should I create a grid of all the possible combinations, should the shapes be updated synchronously?

It's a bit of a pity to reason about this still with the old spec in mind. In the new one as I understood Fynn one would go by reference axes... And I suppose yes, a (hyper-)grid would be the correct solution. But for some reason that sounds too complicated as a result. Poof, this needs more thinking...
